src/Hadrian/Haskell/Cabal/Parse.hs

```diff
@@ -8,36 +8,287 @@
 import qualified Distribution.Types.CondTree            as C
+import qualified Distribution.Types.MungedPackageId    as C (mungedName)
 import qualified Distribution.Verbosity                 as C
 
-import Hadrian.Haskell.Cabal.Type
-
--- | Parse a Cabal file.
-parseCabal :: FilePath -> IO Cabal
-parseCabal file = do
-    gpd <- liftIO $ C.readGenericPackageDescription C.silent file
-    let pd      = C.packageDescription gpd
-        pkgId   = C.package pd
-        name    = C.unPackageName (C.pkgName pkgId)
-        version = C.display (C.pkgVersion pkgId)
-        libDeps = collectDeps (C.condLibrary gpd)
-        exeDeps = map (collectDeps . Just . snd) (C.condExecutables gpd)
-        allDeps = concat (libDeps : exeDeps)
-        sorted  = sort [ C.unPackageName p | C.Dependency p _ <- allDeps ]
-        deps    = nubOrd sorted \\ [name]
-    return $ Cabal deps name (C.synopsis pd) version
-
-collectDeps :: Maybe (C.CondTree v [C.Dependency] a) -> [C.Dependency]
-collectDeps Nothing = []
-collectDeps (Just (C.CondNode _ deps ifs)) = deps ++ concatMap f ifs
-  where
-    f (C.CondBranch _ t mt) = collectDeps (Just t) ++ collectDeps mt
+import Base
+import Builder hiding (Builder)
+import Context -- .Type
+import Flavour (args)
+import GHC.Packages (rts)
+import Hadrian.Expression
+import Hadrian.Haskell.Cabal.Type ( Cabal( Cabal ) )
+import Hadrian.Haskell.Cabal.Configured
+import Hadrian.Oracles.TextFile
+import Hadrian.Target
+import Settings
+import Oracles.Setting
+
+parseCabalPkgId :: FilePath -> IO String
+parseCabalPkgId file = C.display . C.package . C.packageDescription <$> C.readGenericPackageDescription C.silent file
+
+
+biModules :: C.PackageDescription -> (C.BuildInfo, [ModuleName.ModuleName])
+biModules pd = go [ comp | comp@(bi,_) <- (map libBiModules . maybeToList $ C.library pd)
+                                         ++ (map exeBiModules $ C.executables pd)
+                        , C.buildable bi ]
+  where libBiModules lib = (C.libBuildInfo lib, C.explicitLibModules lib)
+        exeBiModules exe = (C.buildInfo exe
+                           , if isHaskell (C.modulePath exe) -- if "main-is: ..." is not a .hs or .lhs file, do
+                                                             -- not inject "Main" into the modules.  This does
+                                                             -- not respect "-main-is" ghc-arguments!  See GHC.hs
+                                                             -- in Distribution.Simple.GHC from Cabal for the glory
+                                                             -- details.
+                             then ModuleName.main : C.exeModules exe
+                             else C.exeModules exe)
+        go [] = error "no buildable component found"
+        go [x] = x
+        go _  = error "can not handle more than one buildinfo yet!"
+        isHaskell fp = takeExtension fp `elem` [".hs", ".lhs"]
+
+parseCabal :: Context -> Action Cabal
```

Comment:
Please add comments to all exported functions.

Reply:
Fixed in an upcoming commit.